frontend/device/gesturedetector.lua

```diff
@@ -253,50 +276,56 @@ function GestureDetector:switchState(state_new, tev, param)
 end
 
 function GestureDetector:clearState(slot)
+    print("GestureDetector:clearState", slot)
     self.states[slot] = self.initialState
-    self.hold_timer_id[slot] = nil
+    self.pending_hold_timer[slot] = nil
     self.detectings[slot] = false
     self.first_tevs[slot] = nil
     self.last_tevs[slot] = nil
     self.multiswipe_directions = {}
     self.multiswipe_type = nil
+
+    -- Also clear any pending hold callbacks on that slot.
+    -- (single taps call this, so we can't clear double_tap callbacks without being caught in an obvious catch-22 ;)).
+    self.input:clearTimeout(slot, "hold")
 end
 
 function GestureDetector:setNewInterval(type, interval)
     if type == "ges_tap_interval" then
-        ges_tap_interval = interval
+        ges_tap_interval = TimeVal:new{ usec = interval }
     elseif type == "ges_double_tap_interval" then
-        ges_double_tap_interval = interval
+        ges_double_tap_interval = TimeVal:new{ usec = interval }
     elseif type == "ges_two_finger_tap_duration" then
-        ges_two_finger_tap_duration = interval
+        ges_two_finger_tap_duration = TimeVal:new{ usec = interval }
     elseif type == "ges_hold_interval" then
-        ges_hold_interval = interval
+        ges_hold_interval = TimeVal:new{ usec = interval }
     elseif type == "ges_pan_delayed_interval" then
-        ges_pan_delayed_interval = interval
+        ges_pan_delayed_interval = TimeVal:new{ usec = interval }
     elseif type == "ges_swipe_interval" then
-        ges_swipe_interval = interval
+        ges_swipe_interval = TimeVal:new{ usec = interval }
     end
 end
 
 function GestureDetector:getInterval(type)
     if type == "ges_tap_interval" then
-        return ges_tap_interval
+        return ges_tap_interval:tousecs()
     elseif type == "ges_double_tap_interval" then
-        return ges_double_tap_interval
+        return ges_double_tap_interval:tousecs()
     elseif type == "ges_two_finger_tap_duration" then
-        return ges_two_finger_tap_duration
+        return ges_two_finger_tap_duration:tousecs()
     elseif type == "ges_hold_interval" then
-        return ges_hold_interval
+        return ges_hold_interval:tousecs()
     elseif type == "ges_pan_delayed_interval" then
-        return ges_pan_delayed_interval
+        return ges_pan_delayed_interval:tousecs()
     elseif type == "ges_swipe_interval" then
-        return ges_swipe_interval
+        return ges_swipe_interval:tousecs()
     end
 end
 
 function GestureDetector:clearStates()
-    self:clearState(0)
-    self:clearState(1)
+    for k, _ in pairs(self.states) do
```

Comment:
~~pairs or ipairs?~~ ok, 0 is not Lua. So, I guess pairs works when we set numbers 0 and 1.

Reply:
Yep, it's a hash ;).